Link against the C++ standard library when C++ interop is enabled

martinboehme opened the pull request:
This doesn't yet allow including C++ headers on platforms where libc++ isn't the default; see comments in UnixToolChains.cpp for details.

However, it does, for example, allow throwing and catching exceptions in C++ code used through interop, unblocking https://github.com/apple/swift/pull/30674.

Part of SR-12469

martinboehme committed:
Link against the C++ standard library when C++ interop is enabled.

This doesn't yet allow including C++ headers on platforms where libc++
isn't the default; see comments in UnixToolChains.cpp for details.

However, it does, for example, allow throwing and catching exceptions in C++
code used through interop, unblocking
https://github.com/apple/swift/pull/30674/files.

martinboehme commented:
@compnerd I'd like your opinion on this in particular as you switched from `clang++` (even without C++ interop enabled) to `clang` in https://github.com/apple/swift/pull/26110.

I think this addresses all of the issues you raise in the code comments. I've rewritten the comment to reflect what the reality is now, but let me know if you think there's more I should add.

gribozavr commented:
@swift-ci Please smoke test

MForster commented on lib/Driver/DarwinToolChains.cpp:
```diff
@@ -712,6 +712,11 @@ toolchains::Darwin::constructInvocation(const DynamicLinkJobAction &job,
   Arguments.push_back(context.Args.MakeArgString(getTriple().getArchName()));
 
+  if (context.cxxInteropEnabled()) {
+    // We only support libc++ on Darwin.
```
This is an ambiguous comment. It could mean either of these:

1) We don't support other standard library implementations on Darwin.
2) We don't support other platforms.

I think you mean the former?

martinboehme replied:
Yes -- thanks for pointing out the ambiguity! I've changed this to "On Darwin, we only support libc++." WDYT?

MForster replied:
SG

MForster commented on lib/Driver/UnixToolChains.cpp:
```diff
@@ -182,27 +182,30 @@ toolchains::GenericUnix::constructInvocation(const DynamicLinkJobAction &job,
-  // default C++ runtime is `libstdc++` which is unsupported on the target but
-  // as the builds are usually cross-compiled from Linux, libstdc++ is going to
-  // be present.  This results in linking the wrong version of libstdc++
-  // generating invalid binaries.  It is also possible to use different C++
```
I think it would be good if we didn't lose this information about Android-specific problems.

martinboehme replied:
I felt conflicted about removing this as well -- but I'm not sure this is the best place to surface this information. (This is also why I asked @compnerd to take a look.)

This sounds to me almost more like an issue with Clang (or the way it's configured) than with Swift itself. Assuming we're passing the right `-target` to Clang (we're doing that, right?), shouldn't it select the correct `-stdlib` default for the platform? If it isn't, shouldn't we file this as a bug against Clang (and potentially put a warning in user-facing documentation until the bug is fixed) rather than putting this in a code comment?

MForster replied:
Sounds good to me as well. I think we should just not lose that information.

martinboehme replied:
Agree that we shouldn't lose this information.

I think the underlying issue here is that some component is exhibiting host / target confusion -- either Clang, or Swift in the way it uses Clang. Until we know which of the two it is, maybe the best thing is to do here is to open a Swift bug, e.g. "Ensure C++ interop defaults to the correct C++ standard library for the target, not the host".

That bug would reference the comment I'm deleting here. The first action item on the bug would be to work out whether this is an issue in Swift or Clang, then proceed from there.

I think that would be a better place for the information to live. For the purposes of this source file, the goal of the comment I'm deleting is really just to make it clear to maintainers that they shouldn't replace `clang` with `clang++` unconditionally, because that might have undesirable consequences. I've tried to fulfill that purpose with the comment "We don't use `clang++` unconditionally because...".

WDYT?

Let's leave this comment thread open until @compnerd has had a chance to respond.

MForster replied:
SG

martinboehme replied:
> I don't think that there is a good place to surface this information sadly.
> 
> The problem is even more interesting in general: a Linux distribution _can_ have multiple supported C++ runtimes.

That's understood.

> You need to not only ensure that the link is correctly referencing the correct library, but the import is also getting the correct headers as otherwise you have an ABI mismatch.

This (making sure the import is getting the correct headers) is done by passing the appropriate `-stdlib=...` flag to Clang, correct? (Specifically, the Clang used in ClangImporter; the Clang invocation that is used on some platforms to link the executable should receive the same `-stdlib` flag.)

> Passing the correct `-target` doesn't do what you want (and really cannot, it can default, but that doesn't make it correct).

I understand that the default C++ standard library may not be what the user wants -- but shouldn't Clang at least choose a default that is supported on the `-target` that was specified (even if the user may want a different C++ standard library in case the target supports more than one)?

This is what I was proposing opening a bug for. I'm basing this on the following comment in UnixToolChains.cpp:

```
  // [...] This is particularly
  // important for platforms like android where as it is a Linux platform, the
  // default C++ runtime is `libstdc++` which is unsupported on the target but
  // as the builds are usually cross-compiled from Linux, libstdc++ is going to
  // be present.  This results in linking the wrong version of libstdc++
  // generating invalid binaries.
```

IIUC, this is saying the following:

*  Android doesn't support libstdc++; executables compiled for it need to use a different standard library (presumably libc++?).
*  When using `clang++` to cross-compile for Android (target) on Linux (host) without a `-stdlib` option, Clang will choose libstdc++ as the default because this is the default on the host.

If true, this second point seems wrong. IMO, Clang's logic for choosing a default for `-stdlib` should be based on what the target is, not what the host is. Is my expectation here wrong? Are there technical reasons for why the `-stdlib` default cannot be made dependent on what the `-target` is?

> You need to pass `-stdlib=` with the correct value. I may want to build against gnustl or libc++ or even a different C++ runtime. There is no correct answer here, the user is the only one who can know for certain.
> 
> If you really want to lift this, I think that you need to expose the C++ runtime through to the driver and require that the user specifies the runtime that they are trying to use.

I've added an option `-experimental-cxx-stdlib` to the driver that allows the user to specify which standard library they want.

I haven't made this a required flag though; as Clang doesn't require `-stdlib` to be specified (and will choose a hopefully meaningful default if it isn't), I think we shouldn't either?

> I think that clang is actually correct in its behaviour here - its not a configuration issue, its a the user must tell us what they are trying to do.

Agree that the user should have this option; I was just hoping that Clang would choose a useful default (see above).

martinboehme replied:
Update: I've created https://bugs.swift.org/browse/SR-12664 to track the Android issue and capture the information there.

MForster commented on lib/Driver/WindowsToolChains.cpp:
```diff
@@ -86,27 +86,13 @@ toolchains::Windows::constructInvocation(const DynamicLinkJobAction &job,
-  // surface this via a driver flag.  For now, opt for the simpler approach of
-  // just using `clang` and avoid a dependency on the C++ runtime.
-  const char *Clang = "clang";
+  // We use `clang++` if C++ interop is enabled, `clang` otherwise.
```
(nit) This is exactly what the code says :-), so I would argue that the comment is just noise.

martinboehme replied:
Good point -- done.

martinboehme committed:
Address review comments by MForster.

martinboehme committed:
Add driver flags for C++ interop.

The flags (-enable-experimental-cxx-interop and -experimental-cxx-stdlib) carry
"experimental" in the name to emphasize that C++ interop is still an
experimental feature.

martinboehme commented:
I've added a commit that adds experimental driver flags for C++ interop. I think this addresses some of the issues raised -- I'll respond to specific comments in a moment.

martinboehme commented on lib/Driver/DarwinToolChains.cpp:
```diff
@@ -712,6 +712,11 @@ toolchains::Darwin::constructInvocation(const DynamicLinkJobAction &job,
+  if (context.cxxInteropEnabled()) {
+    // On Darwin, we only support libc++.
+    Arguments.push_back("-lc++");
+  }
```
Done.

martinboehme commented on lib/Driver/ToolChain.cpp:
```diff
@@ -67,6 +67,19 @@ ToolChain::JobContext::getTemporaryFilePath(const llvm::Twine &name,
+  }
+
+  return false;
+}
```
Thanks -- I wasn't aware of the convention marking driver flags as "experimental", which conveys exactly what we want here (i.e. warning people not to use this in production).

martinboehme commented on lib/Driver/UnixToolChains.cpp:
```diff
@@ -182,27 +182,28 @@ toolchains::GenericUnix::constructInvocation(const DynamicLinkJobAction &job,
+  // In practice, using libc++ on Linux, for example, does not work because the
+  // SwiftGlibc module definition is incompatible with libc++'s header layout.
+  // We probably need to ensure that we use libc++'s own module map instead of 
+  // the SwiftGlibc module map.
```
Do you mean I need to pass `-fmodules-local-submodule-visibility`?

I've tried that, but it doesn't work -- and I don't really see how it would make a difference, as the error I'm getting is a cyclic dependency between modules `std_config` (from libc++'s module map) and `SwiftGlibc`, and IIUC, submodule visibility shouldn't have an influence on this?

Anyway, I've deleted the whole comment referring to libc++ and SwiftGlibc, so this should no longer be an issue for this PR, but I'm still interested in this problem more generally. I'll post a gist later containing the test I'm running and the output I'm getting, and I'd be interested in your insights.

martinboehme replied:
Update: Here's a gist with a test that shows the problem, as well as the error messages produced by Swift:

https://gist.github.com/martinboehme/9faa7747ba24d29f6f0f6915bb3dcb7f

As the error messages note, there's a cyclic dependency between the modules `std_config` (from libc++) and `SwiftGlibc`.

This cyclic dependency comes about as follows:

*  `/usr/include/c++/v1/__config` (from libc++'s module `std_config`) includes `/usr/include/features.h` (from the module `SwiftGlibc`)
*  `/usr/include/inttypes.h` (from `SwiftGlibc`) includes `<stdint.h>` (which because of include paths resolves to `/usr/include/c++/v1/stdint.h`), which includes `/usr/include/c++/v1/__config` (from`std_config`)

Note that the cyclic dependency between the modules, not on the level of files, i.e. it's not the case that `__config` includes a `SwiftGlibc` header file that then (possibly indirectly) includes `__config` again.

It seems that a solution might be to move `features.h` to a different top-level module outside the `SwiftGlibc` module. `__config` doesn't include much else, so this should break the cycle, and `features.h` isn't a standard C header anyway -- IIUC it's really an implementation detail of glibc, so it would be cleaner in any case not to expose it from `SwiftGlibc`.

WDYT?

compnerd replied:
That is a problem we actually need to solve.  The modules are already highly opaque, and this only makes it more complicated.  Actually, SwiftGlibc should not have any dependencies on `std`, just on `Glibc`.  What you are describing sounds like a failure of the `Glibc` module to fully modularlize the C library and having parts of the C library escape into the newly introduced C++ module.

martinboehme replied:
> That is a problem we actually need to solve.

Agree. @gribozavr and I are working on this and have some ideas, though they're still a work in progress -- but we expect to post a proposal on the forum soon.

> The modules are already highly opaque, and this only makes it more complicated. Actually, SwiftGlibc should not have any dependencies on `std`, just on `Glibc`. What you are describing sounds like a failure of the `Glibc` module to fully modularlize the C library and having parts of the C library escape into the newly introduced C++ module.

Unfortunately, it's more tricky than that. The problem is that, fundamentally, glibc and libc++ aren't layered with respect to each other because glibc inadvertently ends up including libc++'s wrapper headers. See the example I discuss above for `stdint.h.`.

The options are:
*  Modify glibc
    * This seems infeasible, as we have to be compatible with existing distributions.
*  Create a single module containing the glibc _and_ libc++ headers
    * This also seems infeasible, as libc++ already ships with a module map which, again, we can't modify for compatibility reasons.
*  Don't modularize glibc at all; instead, include its headers textually
    * Swift obviously can't include the headers textually; instead, we would build `SwiftGlibc` from a header that wraps the glibc headers.

As I say, we plan to post more details on the forum soon.

MForster replied:
This is getting addressed in https://github.com/apple/swift/pull/32404.

martinboehme commented on lib/Driver/UnixToolChains.cpp:
```diff
@@ -182,27 +182,28 @@ toolchains::GenericUnix::constructInvocation(const DynamicLinkJobAction &job,
+  // to `-Xcc` and `-Xclang-linker`, e.g.
+  // `-Xcc -stdlib=libc++ -Xclang-linker -stdlib=libc++`.
+  // Once there is a driver flag for C++ interop, we will probably also want to
+  // add a driver flag for selecting the C++ standard library.
```
Not sure I understand -- `clang++` is only used when C++ interop is turned on, and `clang` continues to be used otherwise. As C++ interop is still experimental, I was assuming no one uses it outside of experiments, so there's nothing that this could break; is this assumption wrong?

At any rate, there is now also an experimental driver flag for selecting the C++ standard library (which does the equivalent of the `-Xcc -stdlib=libc++ -Xclang-linker -stdlib=libc++` in the comment above).

martinboehme commented on lib/Driver/UnixToolChains.cpp:
```diff
@@ -182,27 +182,28 @@ toolchains::GenericUnix::constructInvocation(const DynamicLinkJobAction &job,
+  // SwiftGlibc module definition is incompatible with libc++'s header layout.
+  // We probably need to ensure that we use libc++'s own module map instead of 
+  // the SwiftGlibc module map.
+  const char *Clang = context.cxxInteropEnabled()? "clang++" : "clang";
```
Done.

compnerd commented on lib/Driver/WindowsToolChains.cpp:
```diff
@@ -85,28 +85,14 @@ toolchains::Windows::constructInvocation(const DynamicLinkJobAction &job,
-  // surface this via a driver flag.  For now, opt for the simpler approach of
-  // just using `clang` and avoid a dependency on the C++ runtime.
-  const char *Clang = "clang";
+  const char *Clang =
```
I think that this might want to use `clang-cl` to get all the additional defines that are needed for the Windows C++ runtime.  Selecting the driver could be a shared function in `Toolchain`.  Similar to Unix, could you rename this to `driver`?

martinboehme replied:
> I think that this might want to use `clang-cl` to get all the additional defines that are needed for the Windows C++ runtime.

Do you mean just for the C++ case or also for C? (`clang-cl` sounds like it's the C variant, i.e. analogous to `clang`; it there also a C++ variant of `clang-cl`?)

> Selecting the driver could be a shared function in `Toolchain`.

I understand the sentiment, but I don't really see a good way to do this.

Presumably, this would be some non-virtual function called, maybe, `getClangLinkerDriver()`? But if we want it to return `clang` on Unix and `clang-cl` on Windows, how would we get it do to that?

We could instead make the function virtual and provide different implementations for Unix and Windows, but now we have the same amount of code duplication that we have currently, and we have the additional question of how to implement this function for Darwin (which doesn't even use `clang` to link).

Am I misunderstanding? Did you have something different in mind?

> Similar to Unix, could you rename this to `driver`?

Done (renamed to `LinkerDriver`).

compnerd replied:
Switching between `clang` and `clang-cl` would be based on `T.isOSWindows()` where `T` is `llvm::Triple`.  The Darwin case is something which needs to be addressed (both Jordan and I agreed on this but neither one of us got around to fixing :-() - it should switch over to `clang` as the linker driver, the original reasons for embedding the logic no longer apply as clang is so intricately tied into Swift.

martinboehme replied:
> Switching between `clang` and `clang-cl` would be based on `T.isOSWindows()` where `T` is `llvm::Triple`.

Shouldn't this be `--driver_mode=cl` rather than `clang-cl`? (See discussion above.)

This also makes this something that is easier to implement, because it's just something that can be added to `Arguments` locally in `WindowsToolChains.cpp` (i.e. the function that finds the driver doesn't need to inspect the triple).

> The Darwin case is something which needs to be addressed (both Jordan and I agreed on this but neither one of us got around to fixing :-() - it should switch over to `clang` as the linker driver, the original reasons for embedding the logic no longer apply as clang is so intricately tied into Swift.

Ah, got it.

For the time being, though, all I've done is to pull out the code for finding the linker driver into a common function in `ToolChain`. I'd like to keep this particular PR a behavior-preserving change. Followup PRs will then add `--driver_mode=cl` on Windows and switch Darwin to use Clang as the linker driver.

This will isolate the behavior changes into smaller PRs, which makes it easier to revert them if it turns out they have undesired consequences. WDYT?

compnerd replied:
Sounds good to me.

compnerd commented on test/Driver/cxx_interop.swift:
```diff
@@ -0,0 +1,10 @@
+
+// STDLIB: swift
+// STDLIB-DAG: -enable-cxx-interop
+// STDLIB-DAG: -Xcc -stdlib=libc++
```
Could you add a negative check on no `-experimental-cxx-stdlib` being passed?  Id like to ensure that the MSVCPRT path is not broken.

martinboehme replied:
Done (and discovered what I think is a `FileCheck` bug in the process).

martinboehme committed:
Merge branch 'master' into cxx-link-stdlib

martinboehme committed:
Respond to review comments from @compnerd.

martinboehme commented on lib/Driver/DarwinToolChains.cpp:
```diff
@@ -855,6 +860,13 @@ toolchains::Darwin::validateArguments(DiagnosticEngine &diags,
+
+  // If a C++ standard library is specified, it has to be libc++.
+  if (auto arg = args.getLastArg(options::OPT_experimental_cxx_stdlib)) {
+    if (StringRef(arg->getValue()) != "libc++") {
```
Not sure what you mean? I don't have a `std::string` here yet -- `arg->getValue()` returns a `const char *`. Creating a string from that would require a memory allocation -- using a `StringRef` instead avoids that.

compnerd replied:
For some reason I thought that there was `int (const char *, const char *)` overload that allowed you to safely get to `strncmp`.  I just dislike the `StringRef` construction since it tends to be a bit tricky, but sure.

martinboehme replied:
Fine to mark this resolved then?

martinboehme commented on lib/Driver/UnixToolChains.cpp:
```diff
@@ -182,27 +182,18 @@ toolchains::GenericUnix::constructInvocation(const DynamicLinkJobAction &job,
+  // is desired.
+  const char *Clang =
+      context.Args.hasArg(options::OPT_enable_experimental_cxx_interop) ?
+      "clang++" : "clang";
```
See comment below -- I don't see a good way to do this.

martinboehme replied:
Update: Done (see discussion below).

martinboehme commented on lib/Driver/UnixToolChains.cpp:
```diff
@@ -182,27 +182,18 @@ toolchains::GenericUnix::constructInvocation(const DynamicLinkJobAction &job,
+  // a C++ standard library if it's not needed, in particular because the
+  // standard library that `clang++` selects by default may not be the one that
+  // is desired.
+  const char *Clang =
```
Done -- but I've called it `LinkerDriver` to prevent confusion with the Swift driver.

gribozavr commented on include/swift/Option/Options.td:
```diff
@@ -519,6 +519,14 @@ def enable_experimental_concise_pound_file : Flag<["-"],
 
+def enable_experimental_cxx_interop :
+  Flag<["-"], "enable-experimental-cxx-interop">,
+  HelpText<"Enable C++ interop code generation and config directives">;
```
```suggestion
  HelpText<"Allow importing C++ modules into Swift (experimental feature)">;
```

I'm not sure what config directives are.

martinboehme replied:
I copied the `HelpText` from the frontend flag `-enable-cxx-interop` -- but I like your wording better. Done.

gribozavr commented:
@swift-ci Please test

martinboehme committed:
Factor out logic for finding the linker driver into a common function.

martinboehme committed:
Change help text for -enable-experimental-cxx-interop.

martinboehme commented:
@compnerd Your rewiew state is showing as "Changes requested", but I can't get github to show me those changes. Maybe those requested changes were for an outdated part of the PR?

Anything else that should be changed before this PR is ready to be merged?

gribozavr commented:
@swift-ci Please test

gribozavr commented:
@swift-ci Please test Windows

compnerd commented on lib/Driver/ToolChains.cpp:
```diff
@@ -166,6 +166,17 @@ void ToolChain::addCommonFrontendArgs(const OutputInfo &OI,
+    arguments.push_back("-enable-cxx-interop");
+  }
+  if (const Arg *arg =
+      inputArgs.getLastArg(options::OPT_experimental_cxx_stdlib)) {
```
This seems oddly formatted, `clang-format` please?

martinboehme replied:
Done.

(The formatting that clang-format applied also seems a bit strange to me -- but who am I to argue with clang-format?)

compnerd commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1321,6 +1332,26 @@ void ToolChain::getRuntimeLibraryPaths(SmallVectorImpl<std::string> &runtimeLibP
+  // is desired.
+  const char *LinkerDriver =
+      Args.hasArg(options::OPT_enable_experimental_cxx_interop) ?
+      "clang++" : "clang";
```
`clang-format`?

martinboehme replied:
Done.

compnerd commented on lib/Driver/UnixToolChains.cpp:
```diff
@@ -307,6 +285,13 @@ toolchains::GenericUnix::constructInvocation(const DynamicLinkJobAction &job,
 
+  // Link against the desired C++ standard library.
+  if (const Arg *A =
+      context.Args.getLastArg(options::OPT_experimental_cxx_stdlib)) {
```
`clang-format` please

martinboehme replied:
Done.

martinboehme committed:
Apply clang-format.

martinboehme commented:
@swift-ci Please test

gribozavr commented:
@swift-ci Please test

gribozavr commented:
@swift-ci Please test Windows

martinboehme committed:
Merge branch 'master' into cxx-link-stdlib

martinboehme commented:
@swift-ci Please test

martinboehme commented:
@swift-ci Please test Windows

gribozavr commented:
@swift-ci Please test macOS

gribozavr commented:
@swift-ci Please test Windows

compnerd reviewed (approved):
Please add a test case to ensure that msvcprt demonstrating can be used by default on Windows.

MForster committed:
Merge branch 'master' into cxx-link-stdlib

MForster commented:
@swift-ci Please test

MForster commented:
> Please add a test case to ensure that msvcprt demonstrating can be used by default on Windows.

I've filed SR-13164 for this:

MForster commented:
@swift-ci Please test OS X Platform

MForster commented:
@swift-ci Please test OS X Platform

MForster committed:
Specify -sdk "" for linker tests.

This is necessary after commit
1be17a255029b4fd6a2e848baf6354687fbcc76a

MForster commented:
@swift-ci Please test

MForster commented:
@swift-ci Please test
